File: examples/cookbook/05_parallel_fanout.py
"""Parallel FanOut"""

# --- NATIVE ---
from google.adk.agents.llm_agent import LlmAgent
from google.adk.agents.parallel_agent import ParallelAgent

fanout_native = ParallelAgent(
    name="parallel_search",
    sub_agents=[
        LlmAgent(name="web", model="gemini-2.5-flash", instruction="Search web."),
        LlmAgent(name="db", model="gemini-2.5-flash", instruction="Search database."),
    ],
)

# --- FLUENT ---
from adk_fluent import Agent, FanOut

fanout_fluent = (
    FanOut("parallel_search")
    .branch(Agent("web").model("gemini-2.5-flash").instruct("Search web."))
    .branch(Agent("db").model("gemini-2.5-flash").instruct("Search database."))
    .build()
)

# --- ASSERT ---
assert type(fanout_native) == type(fanout_fluent)
assert len(fanout_fluent.sub_agents) == 2

File: ts/examples/cookbook/05_parallel_fanout.ts
/**
 * 05 — Parallel FanOut
 *
 * Multi-source research: web, papers, and news searched in parallel.
 *
 * Demonstrates the `FanOut` builder which wraps `ParallelAgent`. Each branch
 * runs concurrently. The same workflow can also be expressed via the
 * `.parallel()` method on any agent (Python's `|` operator equivalent).
 */
import assert from "node:assert/strict";
import { Agent, FanOut } from "../../src/index.js";

const research = new FanOut("research")
  .branch(new Agent("web", "gemini-2.5-flash").instruct("Search the web.").writes("web_results"))
  .branch(new Agent("papers", "gemini-2.5-flash").instruct("Search arXiv.").writes("paper_results"))
  .branch(new Agent("news", "gemini-2.5-flash").instruct("Search news.").writes("news_results"))
  .build() as Record<string, unknown>;

assert.equal(research._type, "ParallelAgent");
assert.equal((research.subAgents as unknown[]).length, 3);

// Equivalent expression using `.parallel()` (the TS analog of Python's `|`).
const research2 = new Agent("web", "gemini-2.5-flash")
  .instruct("Search the web.")
  .parallel(new Agent("papers", "gemini-2.5-flash").instruct("Search arXiv."))
  .parallel(new Agent("news", "gemini-2.5-flash").instruct("Search news."))
  .build() as Record<string, unknown>;

assert.equal(research2._type, "ParallelAgent");
assert.equal((research2.subAgents as unknown[]).length, 3);

export { research, research2 };
